lib/ecto/query/planner.ex

```diff
@@ -2280,6 +2280,39 @@ defmodule Ecto.Query.Planner do
     end)
   end
 
+  defp handle_subquery_struct(select, requested_fields, ix, query) do
+    available_fields = subquery_source_fields(select)
+    requested_fields = List.wrap(requested_fields)
+
+    Enum.each(requested_fields, fn field ->
+      unless field in available_fields do
+        error!(query, "field `#{field}` in struct/2 is not available in the subquery. " <>
+                     "Subquery only returns fields: #{inspect(available_fields)}")
+      end
+    end)
+
+    schema =
+      case select do
+        {:struct, schema, _} -> schema
+        {:source, {_, schema}, _, _} when not is_nil(schema) -> schema
+        _ ->
+          error!(query, "it is not possible to return a struct subset of a subquery that does not return a schema or a struct")
+      end
+
+    types =
+      Enum.map(requested_fields, fn field ->
+        {:ok, type} = subquery_type_for(select, field) 
+        {field, type}
+      end)
+
+    field_exprs =
+      Enum.map(types, fn {field, type} ->
+        {{:., [type: type], [{:&, [], [ix]}, field]}, [], []}
```

Comment:
We should probably use `select_field` to build this information. In particular, we don't attach the type information anywhere in the other `take!` calls, so perhaps we shouldn't do so here.